### DISCUSSION TITLE
Clone spreadsheet loose drawings

### DISCUSSION THREAD

**Discussion by PasCapableTrouverUnNom** [PHPOffice/PhpSpreadsheet#2663] [Q&A]
I know this was an issue that has been solved from previous versions (Used to work on 1.14 but not anymore in 1.21)

Summary: Cloning a worksheet loose drawing (in 1.14 we used to lost only parameters)

       //open the final spreadsheet
        $spreadsheetFinale=new Spreadsheet();
        $NomFichierFinal = storage_path('app/Output/Rapport_' . $userId . '.xlsx');
       //I open multiple spreadsheets ($exportSelected is that list) and add their sheet to the new file
        foreach ($exportSelected as $selected) {
            $NomFichier = storage_path('app/Output/FicheBrute_' . $userId . '_' . $selected['reporttype'] . '_' . $selected['ObjID'] . '.xlsx');
            //Open the Model
            $reader = new reader\Xlsx();
            $spreadsheet = $reader->load($NomFichier);
            $sheetNames = $spreadsheet->getSheetNames();
           //For each sheet of the  $NomFichier I clone the sheet in the new file $SpreadsheetFinale
            foreach ($sheetNames as $sheet) {
               //Select the sheet
                $OutputSheet = $sheet . '_' . $selected['reporttype'] . '_' . $selected['ObjID'];
               //clone the sheet
                $cloneWorksheet = clone $spreadsheet->getSheetByName($sheet);
               //add the sheet to the new file
                $spreadsheetFinale->addExternalSheet($cloneWorksheet);
                //Change the name of the worksheet in the new spreadsheet
                $sheet = $spreadsheetFinale->setActiveSheetIndexByName($sheet);
                $sheet->setTitle($OutputSheet);
            }
        }
       //Once all worksheets has been added we delete the worksheet named 'Worksheet' which is default when we create a spreadsheet
        $spreadsheetFinale->setActiveSheetIndexByName('Worksheet');
        $sheetIndex = $spreadsheetFinale->getActiveSheetIndex();
        $spreadsheetFinale->removeSheetByIndex($sheetIndex);
          //Create a new writer to write the new spreadsheet
        $writerFinal = new writer\Xlsx($spreadsheetFinale);
        $writerFinal->save($NomFichierFinal);

        return response()->download($NomFichierFinal);

Note:
- I tried setIncludeChart(True);, the result is a corrupt file
- This exact code (I removed the input for sake of simplicity) works just fine in 1.14
- The spreadsheet selected have been created by PhpSpreadsheet in a previous step (the user see them all, select and then we compose a compilaiton with the selected spreadsheet).   The output of this first step (the excel files) are PERFECT when I look at them individually.

Any hint or path would be appreciated... I saw some similar problem in very old post but nothing recent and the solutions do not seems to work.

Thanks

**Comment by oleibman** [comment]
The following code, based substantially on yours, should do what you want.
```php
        //open the final spreadsheet
        $spreadsheetFinale = new Spreadsheet();
        $exportSelected = [
            'discussion.2663.file1.xlsx',
            'discussion.2663.file2.xlsx',
        ];
        //I open multiple spreadsheets ($exportSelected is that list) and add their sheet to the new file
        $fileNumber = 0;
        foreach ($exportSelected as $selected) {
            ++$fileNumber;
            //Open the Model
            $reader = new XlsxReader();
            $spreadsheet = $reader->load($selected);
            //For each sheet of the  $NomFichier I clone the sheet in the new file $SpreadsheetFinale
            foreach ($spreadsheet->getAllSheets() as $sheet) {
                //Select the sheet
                $title = $sheet->getTitle();
                $OutputSheet = "file$fileNumber" . '_' . $title;
                $newWorksheet = $spreadsheet->duplicateWorksheetByTitle($title);
                //add the sheet to the new file
                $ws = $spreadsheetFinale->addExternalSheet($newWorksheet);
                //Change the name of the worksheet in the new spreadsheet
                $ws->setTitle($OutputSheet);
            }
            $spreadsheet->disconnectWorksheets();
        }
        //Once all worksheets has been added we delete the worksheet named 'Worksheet' which is default when we create a spreadsheet
        $spreadsheetFinale->setActiveSheetIndexByName('Worksheet');
        $sheetIndex = $spreadsheetFinale->getActiveSheetIndex();
        $spreadsheetFinale->removeSheetByIndex($sheetIndex);
        //Create a new writer to write the new spreadsheet
        $NomFichierFinal = 'discussion.2663.merged.xlsx';
        $writerFinal = new XlsxWriter($spreadsheetFinale);
        $writerFinal->save($NomFichierFinal);
        echo "saved $NomFichierFinal\n";
```
